src/greenlight/runner.clj

```diff
@@ -77,6 +108,54 @@
     true))
 
 
+(defn- run-group
+  [^ExecutorService exec-pool system tests]
+  {:post [(instance? Reader (first %))
+          (instance? Future (second %))]}
+  (let [reader (PipedReader.)
+        writer (PipedWriter. reader)
+        bindings (get-thread-bindings)
+        group-f (reify Callable
+                  (call [_]
+                    (with-bindings bindings
+                      (binding [*out* writer, *err* writer]
+                        (try
+                          (mapv (partial test/run-test! system) tests)
+                          (finally
+                            (.close writer)))))))]
+    [reader (.submit exec-pool group-f)]))
+
+
+(defn- print-reader
+  [^Reader reader]
+  (with-open [br (BufferedReader. reader)]
+    (loop []
+      (when-let [line (.readLine br)]
+        (println line)
+        (recur)))))
+
+
+(defn- execute-parallel
+  "Run a collection of tests, using an executor pool with `n-threads`"
+  [system tests n-threads]
+  (let [exec-pool (Executors/newFixedThreadPool n-threads)]
+    (try
+      (->> tests
+           (group-by #(or (::test/group %) (gensym)))
+           (map val)
+           (map (partial run-group exec-pool system))
```

Comment:
I don't think that tests in the same linearized group also need to get piped to stdout adjacent to each other. I think it would make the most sense if tests get piped to stdout in the same order they were declared in `runner/main` (regardless of what group they're in). This would require some more async-fu to implement, but would gives the test author some extra control, if for example they know that some of their tests take especially long and don't want to block on those first.

Depending on how much you want to optimize this, you could also keep a queue of tests to pipe to stdout next, and bump things to the top of the queue if they fail in the background, which would surface errors sooner.